pkg/component/worker/utils.go

```diff
@@ -17,62 +17,154 @@ limitations under the License.
-	if err != nil {
-		return fmt.Errorf("failed to parse kubelet bootstrap auth from token: %w", err)
-	}
+	if joinToken == "" {
+		bootstrapKubeconfigPath = filepath.Join(k0sVars.DataDir, "kubelet-bootstrap.conf")
+		if !file.Exists(bootstrapKubeconfigPath) {
+			return fmt.Errorf("neiter regular nor bootstrap kubelet kubeconfig files exist and no join token given; dunno how to make kubelet authenticate to API server")
+		}
 
-	if tokenType := token.GetTokenType(clientCfg); tokenType != "kubelet-bootstrap" {
-		return fmt.Errorf("wrong token type %s, expected type: kubelet-bootstrap", tokenType)
+		var err error
+		bootstrapKubeconfig, err = clientcmd.LoadFromFile(bootstrapKubeconfigPath)
+		if err != nil {
+			return fmt.Errorf("failed to parse kubelet bootstrap kubeconfig from file: %w", err)
+		}
+	} else {
+		kubeconfig, err := token.DecodeJoinToken(joinToken)
+		if err != nil {
+			return fmt.Errorf("failed to decode join token: %w", err)
+		}
+
+		// Load the bootstrap kubeconfig to validate it
+		bootstrapKubeconfig, err = clientcmd.Load(kubeconfig)
+		if err != nil {
+			return fmt.Errorf("failed to parse kubelet bootstrap kubeconfig from join token: %w", err)
+		}
+
+		bootstrapKubeconfigPath, err = writeKubeletBootstrapKubeconfig(kubeconfig)
+		if err != nil {
+			return fmt.Errorf("failed to write kubelet bootstrap kubeconfig: %w", err)
+		}
+		defer func() {
+			if err := os.Remove(bootstrapKubeconfigPath); err != nil && !os.IsNotExist(err) {
+				logrus.WithError(err).Error("Failed to remove kubelet bootstrap kubeconfig file")
+			}
+		}()
+
+		logrus.Debug("Wrote kubelet bootstrap kubeconfig file: ", bootstrapKubeconfigPath)
 	}
 
 	kubeletCAPath := path.Join(k0sVars.CertRootDir, "ca.crt")
 	if !file.Exists(kubeletCAPath) {
 		if err := dir.Init(k0sVars.CertRootDir, constant.CertRootDirMode); err != nil {
 			return fmt.Errorf("failed to initialize directory '%s': %w", k0sVars.CertRootDir, err)
 		}
-		err = file.WriteContentAtomically(kubeletCAPath, clientCfg.Clusters["k0s"].CertificateAuthorityData, constant.CertMode)
+		err := file.WriteContentAtomically(kubeletCAPath, bootstrapKubeconfig.Clusters["k0s"].CertificateAuthorityData, constant.CertMode)
 		if err != nil {
 			return fmt.Errorf("failed to write ca client cert: %w", err)
 		}
 	}
-	err = file.WriteContentAtomically(k0sVars.KubeletBootstrapConfigPath, kubeconfig, constant.CertSecureMode)
+
+	if tokenType := token.GetTokenType(bootstrapKubeconfig); tokenType != "kubelet-bootstrap" {
+		return fmt.Errorf("wrong token type %s, expected type: kubelet-bootstrap", tokenType)
+	}
+
+	certDir := filepath.Join(k0sVars.DataDir, "kubelet", "pki")
+	if err := dir.Init(certDir, constant.DataDirMode); err != nil {
+		return fmt.Errorf("failed to initialize kubelet certificate directory: %w", err)
+	}
+
+	// FIXME: This is not completely on par with kubelet code. The kubelet has
+	// some means of retrieving the node name via the cloud provider, if one is
+	// configured.
```

Comment:
The k0s kubelet gets built without the in-tree cloud providers, so this might not apply, after all.